httpclient5/src/main/java/org/apache/hc/client5/http/auth/BearerToken.java

```diff
@@ -0,0 +1,120 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+package org.apache.hc.client5.http.auth;
+
+import java.io.Serializable;
+import java.security.Principal;
+import java.util.Objects;
+
+import org.apache.hc.core5.annotation.Contract;
+import org.apache.hc.core5.annotation.ThreadingBehavior;
+import org.apache.hc.core5.util.Args;
+
+/**
+ * Opaque token {@link Credentials} usually representing a set of claims, often encrypted
+ * or signed. The JWT (JSON Web Token) is among most widely used tokens used at the time
+ * of writing.
+ *
+ * @since 5.3
+ */
+@Contract(threading = ThreadingBehavior.IMMUTABLE)
+public class BearerToken implements Credentials, Serializable {
+
+    private final Principal principal;
+    private final String token;
+
+    /**
+     * The constructor with the user principal and token arguments.
+     *
+     * @param principal the user principal
+     * @param token the token
+     *
+     * @see BasicUserPrincipal
+     * @see NTUserPrincipal
+     */
+    public BearerToken(final Principal principal, final String token) {
+        super();
+        this.principal = principal;
+        this.token = Args.notBlank(token, "Token");
+    }
+
+    /**
+     * The constructor with the username and token arguments.
+     *
+     * @param userName the username
+     * @param token the token
+     */
+    public BearerToken(final String userName, final String token) {
+        this(userName != null ? new BasicUserPrincipal(userName) : null, token);
+    }
+
+    @Override
+    public Principal getUserPrincipal() {
+        return this.principal;
+    }
+
+    public String getUserName() {
+        return this.principal.getName();
+    }
+
+    /**
+     * @deprecated Do not use.
+     */
+    @Deprecated
+    @Override
+    public char[] getPassword() {
+        return null;
+    }
+
+    public String getToken() {
+        return token;
+    }
+
+    @Override
+    public int hashCode() {
+        return this.principal.hashCode();
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o instanceof BearerToken) {
+            final BearerToken that = (BearerToken) o;
+            return Objects.equals(this.principal, that.principal);
```

Comment:
Do you really consider both equal when the access tokens differ, but principal matches?

Reply:
@michael-o I do. This is why I think null principal is sully, but so be it.

Reply:
I can imagine the following: two clients (principals) have been granted to the same token to access a resource on behalf of the resource owner. How should equals behave here?